dipy/reconst/mapmri.py

```diff
@@ -0,0 +1,655 @@
+        >>> from dipy.reconst.mapmri import MapmriModel
+        >>> radial_order = 4
+        >>> map_model = MapmriModel(gtab, radial_order=radial_order)
+        >>> mapfit = map_model.fit(data)
+        >>> odf= mapfit.odf(sphere)
+        """
+
+        self.bvals = gtab.bvals
+        self.bvecs = gtab.bvecs
+        self.gtab = gtab
+        self.radial_order = radial_order
+        self.lambd = lambd
+        self.eap_cons = eap_cons
+        if self.eap_cons:
+            if not have_cvxopt:
+                raise ValueError(
+                    'CVXOPT package needed to enforce constraints')
+            import cvxopt.solvers
+
+        self.anisotropic_scaling = anisotropic_scaling
+        if (gtab.big_delta is None) or (gtab.small_delta is None):
+            self.tau = 1 / (4 * np.pi ** 2)
+        else:
+            self.tau = gtab.big_delta - gtab.small_delta / 3.0
+        self.eigenvalue_threshold = eigenvalue_threshold
+
+        self.ind = self.gtab.bvals <= bmax_threshold
+        gtab_dti = gradient_table(
+            self.gtab.bvals[self.ind], self.gtab.bvecs[self.ind, :])
+        self.tenmodel = dti.TensorModel(gtab_dti)
+        self.ind_mat = mapmri_index_matrix(self.radial_order)
+        self.Bm = b_mat(self.ind_mat)
+
+    @multi_voxel_fit
+    def fit(self, data):
+
+        tenfit = self.tenmodel.fit(data[..., self.ind])
+        evals = tenfit.evals
+        R = tenfit.evecs
+        ind_evals = np.argsort(evals)[::-1]
+        evals = evals[ind_evals]
+        R = R[ind_evals, :]
+        evals = np.clip(evals, self.eigenvalue_threshold, evals.max())
+        if self.anisotropic_scaling:
+            mu = np.sqrt(evals * 2 * self.tau)
+
+        else:
+            mumean = np.sqrt(evals.mean() * 2 * self.tau)
+            mu = np.array([mumean, mumean, mumean])
+
+        qvals = np.sqrt(self.gtab.bvals / self.tau) / (2 * np.pi)
+        qvecs = np.dot(self.gtab.bvecs, R)
+        q = qvecs * qvals[:, None]
+        M = mapmri_phi_matrix(self.radial_order, mu, q.T)
+        # This is a simple empirical regularization, to be replaced
+        I = np.diag(self.ind_mat.sum(1) ** 2)
+        if self.eap_cons:
+            rmax = 2 * np.sqrt(10 * evals.max() * self.tau)
+            r_index, r_grad = create_rspace(11, rmax)
+            K = mapmri_psi_matrix(
+                self.radial_order,  mu, r_grad[0:len(r_grad) / 2, :])
+
+            Q = cvxopt.matrix(np.dot(M.T, M) + self.lambd * I)
+            p = cvxopt.matrix(-1 * np.dot(M.T, data))
+            G = cvxopt.matrix(-1 * K)
+            h = cvxopt.matrix(np.zeros((K.shape[0])), (K.shape[0], 1))
+            cvxopt.solvers.options['show_progress'] = False
+            sol = cvxopt.solvers.qp(Q, p, G, h)
+            if sol['status'] != 'optimal':
+                warn('Optimization did not find a solution')
+
+            coef = np.array(sol['x'])[:, 0]
+        else:
+            pseudoInv = np.dot(
```

Comment:
you can use directly np.linalg.pinv to compute the pseudo inverse instead of building and inverting it, it's more stable than doing directly the matrix inversion yourself.